File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();        
        int left = 0, right = 0;
        int longestSubstringLength = Integer.MIN_VALUE;

        while (right < s.length()) {
            char currentChar = s.charAt(right);

            if (!set.contains(currentChar)) {
                set.add(currentChar);
                right++;
                longestSubstringLength = Math.max(longestSubstringLength, set.size());
            } else {
                set.remove(s.charAt(left));
                left++; 
            }
        }
        
        return longestSubstringLength < 0 ? 0 : longestSubstringLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i = 0 
        ele = set()
        max_window = 0

        for j in range(len(s)):
            while i < j and s[j] in ele:
                ele.remove(s[i])
                i += 1
                
            ele.add(s[j])
            max_window = max(max_window, len(ele))
        
        return max_window